File: sdk-py/getpatter/observability/metric_types.py
"""Typed metric data models for Patter pipeline events.

Shapes mirror LiveKit Agents ``livekit.agents.metrics`` for cross-SDK
compatibility. All models are frozen dataclasses (no pydantic dependency).
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field


@dataclass(frozen=True)
class Metadata:
    """Provider / model metadata attached to a metric."""

    model_name: str | None = None
    model_provider: str | None = None


@dataclass(frozen=True)
class CachedTokenDetails:
    """Breakdown of cached tokens by modality."""

    audio_tokens: int = 0
    text_tokens: int = 0
    image_tokens: int = 0


@dataclass(frozen=True)
class InputTokenDetails:
    """Detailed input token breakdown (mirrors OpenAI / LiveKit shapes)."""

    audio_tokens: int = 0
    text_tokens: int = 0
    image_tokens: int = 0
    cached_tokens: int = 0
    cached_tokens_details: CachedTokenDetails | None = None


@dataclass(frozen=True)
class OutputTokenDetails:
    """Detailed output token breakdown."""

    text_tokens: int = 0
    audio_tokens: int = 0
    image_tokens: int = 0


@dataclass(frozen=True)
class LLMUsage:
    """LLM token usage for a single response (matches LiveKit ``CompletionUsage``)."""

    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0
    prompt_cached_tokens: int = 0
    cache_creation_tokens: int = 0
    cache_read_tokens: int = 0


@dataclass(frozen=True)
class RealtimeUsage:
    """Token usage for an OpenAI Realtime session."""

    session_duration_seconds: float = 0.0
    tokens_per_second: float = 0.0
    input_token_details: InputTokenDetails = field(default_factory=InputTokenDetails)
    output_token_details: OutputTokenDetails = field(default_factory=OutputTokenDetails)
    metadata: Metadata | None = None


@dataclass(frozen=True)
class EOUMetrics:
    """End-of-utterance timing metrics (LiveKit Pattern A).

    All delay fields are in **seconds**.
    """

    end_of_utterance_delay: float
    transcription_delay: float
    on_user_turn_completed_delay: float
    timestamp: float = field(default_factory=time.time)
    speech_id: str | None = None
    metadata: Metadata | None = None


@dataclass(frozen=True)
class InterruptionMetrics:
    """Barge-in / overlap metrics (LiveKit Pattern B, no ML classification).

    All duration fields are in **seconds**.
    """

    total_duration: float
    detection_delay: float
    num_interruptions: int
    num_backchannels: int
    timestamp: float = field(default_factory=time.time)
    prediction_duration: float = 0.0
    metadata: Metadata | None = None


@dataclass(frozen=True)
class TTFBMetrics:
    """Time-to-first-byte latency for a pipeline stage (mirrors Pipecat)."""

    processor: str
    value: float
    timestamp: float = field(default_factory=time.time)
    model: str | None = None


@dataclass(frozen=True)
class ProcessingMetrics:
    """Total processing time for a pipeline stage."""

    processor: str
    value: float
    timestamp: float = field(default_factory=time.time)
    model: str | None = None

File: sdk-ts/src/observability/metric-types.ts
/**
 * Typed metric payload shapes for Patter observability events.
 *
 * These interfaces mirror the Python dataclasses in
 * ``sdk-py/getpatter/observability/metric_types.py`` and are emitted via
 * ``EventBus`` from ``CallMetricsAccumulator``.
 */

export interface Metadata {
  modelName?: string | null;
  modelProvider?: string | null;
}

// ---- LLM usage ----

export interface LLMUsage {
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
  promptCachedTokens?: number;
  cacheCreationTokens?: number;
  cacheReadTokens?: number;
}

// ---- Realtime usage ----

export interface CachedTokenDetails {
  audioTokens: number;
  textTokens: number;
  imageTokens: number;
}

export interface InputTokenDetails {
  audioTokens: number;
  textTokens: number;
  imageTokens: number;
  cachedTokens: number;
  cachedTokensDetails?: CachedTokenDetails | null;
}

export interface OutputTokenDetails {
  textTokens: number;
  audioTokens: number;
  imageTokens: number;
}

export interface RealtimeUsage {
  sessionDurationSeconds: number;
  tokensPerSecond: number;
  inputTokenDetails: InputTokenDetails;
  outputTokenDetails: OutputTokenDetails;
  metadata?: Metadata | null;
}

// ---- EOU metrics ----

/**
 * End-of-utterance timing breakdown.
 *
 * ``endOfUtteranceDelay``     ms from VAD stop → STT final.
 * ``transcriptionDelay``      ms from VAD stop → turn committed to LLM.
 * ``onUserTurnCompletedDelay`` ms from turn committed → pipeline hook done.
 */
export interface EOUMetrics {
  timestamp: number;
  endOfUtteranceDelay: number;
  transcriptionDelay: number;
  onUserTurnCompletedDelay: number;
  speechId?: string | null;
  metadata?: Metadata | null;
}

// ---- Interruption metrics ----

/**
 * Barge-in / interruption measurement.
 *
 * ``predictionDuration`` is always 0 in the simplified no-ML implementation;
 * it is reserved for a future ML-based overlap classifier.
 */
export interface InterruptionMetrics {
  timestamp: number;
  totalDuration: number;
  predictionDuration: number;
  detectionDelay: number;
  numInterruptions: number;
  numBackchannels: number;
  metadata?: Metadata | null;
}

// ---- TTFB / processing metrics ----

export interface TTFBMetrics {
  timestamp: number;
  processor: string;
  model?: string | null;
  value: number;
}

export interface ProcessingMetrics {
  timestamp: number;
  processor: string;
  model?: string | null;
  value: number;
}
